### ISSUE TITLE
eqjoin.group.count is slow in Windows when running rethinkdb as PROXY

### ISSUE THREAD

**Issue by bearxu83** [rethinkdb/rethinkdb#5735]
Rethinkdb 2.3
Windows 10

Running these aggregation functions when the rethinkdb was running in Windows as proxy is much slower than connecting with the cluster which runs in Linux directly.

**Comment by danielmewes** [comment]
@bearxu83 Thanks for the report. To make it easier for us to reproduce and test this, would you mind providing a few additional details?
* What is the hardware on the Windows and Linux servers?
* What is the network connection between the Windows proxy and the Linux server?
* How does your data look? Can you share a typical example document from each involved table?
* What is the exact query you're running?

**Comment by danielmewes** [comment]
Also maybe a stupid question, but have you run the same query a few times against each of the two machines to rule out caching effects?

**Comment by bearxu83** [comment]
Hi, Daniel
The Windows Machine is Core i5 4200U, 8G. The Linux Cluster is in a University Network. I will send you an email about the location of the db and the query. I think maybe it is a bug instead of some caching effect because the difference is very large.

**Comment by danielmewes** [comment]
Yeah it's probably not the caching. Just wanted to make sure you ran the query more than once against each server. Thanks for the details on the setup and query.
How fast is the network connection between your local (?) Windows machine and the Linux servers in the university network?

Can you check how much CPU the RethinkDB proxy on Windows is using while the query is running (e.g. using the task manager)?

It might either be an inefficiency in our Windows port, or the query could be slower because a lot of intermediate results need to be transferred over the network to the proxy.

If you have time to try it, it would also be interesting to see what performance you get when running the Linux version of RethinkDB as a proxy in a virtual machine on your local computer (e.g. with VirtualBox or through Vagrant). If the Linux version turns out to be faster, that would strongly indicate that the inefficiency is actually in our Windows port.

**Comment by bearxu83** [comment]
Good idea. I will do it after my final exam on next Monday. :)

**Comment by bearxu83** [comment]
I installed the Linux version of RethinkDB in a Virtualbox. The speed difference between using proxy and connecting directly disappears when I use another dataset. It seems that I need more time to recheck this issue.

**Comment by danielmewes** [comment]
Interesting. Thanks for investigating @bearxu83 . Were you able to test inside the VirtualBox with the same data set as well? Is the difference still there in that case?

**Comment by bearxu83** [comment]
Today I imported about 3GB data into rethinkdb in Virtualbox in bridge network mode.

The query is like:
`
year_word_count = rdb.db("rm").table("rm_sentences2").limit(200000).eq_join('article_id', rdb.db("rm").table("rm_articles")).count().run(conn)
`
The result
`conn = rdb.connect('192.168.1.44', 28015)  ====> 28s connect directly`
`conn_local = rdb.connect('localhost', 28015) ====> 1min13s connect proxy`

Maybe there is some problem with eq_join, I will test other commands and other datasets.

I will also try to run proxy mode in Linux instead of Windows to see if the difference exists.

**Comment by danielmewes** [comment]
My suspicion is still that the difference stems from the network delays when transferring the data from both tables over the network to the proxy. If you connect directly to the remote server, only the final `count` result will be transferred.
